flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -162,6 +169,49 @@ class FIROpConversion : public mlir::OpConversionPattern<FromOp> {
     return rewriter.create<mlir::LLVM::LoadOp>(loc, ty, p);
   }
 
+  mlir::Value
+  loadStrideFromBox(mlir::Location loc, mlir::Value box, unsigned dim,
+                    mlir::ConversionPatternRewriter &rewriter) const {
+    auto idxTy = lowerTy().indexType();
+    auto c0 = genConstantOffset(loc, rewriter, 0);
+    auto c7 = genConstantOffset(loc, rewriter, 7);
+    auto dimValue = genConstantIndex(loc, idxTy, rewriter, dim);
+    return loadFromOffset(loc, box, c0, c7, dimValue, 2, idxTy, rewriter);
+  }
+  /// Read base address from a fir.box. Returned address has type ty.
+  mlir::Value
+  loadBaseAddrFromBox(mlir::Location loc, mlir::Type ty, mlir::Value box,
+                      mlir::ConversionPatternRewriter &rewriter) const {
+    auto c0 = genConstantOffset(loc, rewriter, 0);
+    auto pty = mlir::LLVM::LLVMPointerType::get(unwrap(ty));
```

Comment:
Isn't `ty` already an LLVM pointer type?

Reply:
It is, but we want a double indirection here since it is the address of an address that we are going to load here.